lib/util/glob-util.js

```diff
@@ -117,41 +117,32 @@ function listFilesToProcess(globPatterns, providedOptions) {
 
     /**
      * Executes the linter on a file defined by the `filename`. Skips
-     * unsupported file extensions and any files that are already linted.
+     * any files that are already linted.
      * @param {string} filename The file to be processed
-     * @param {boolean} shouldWarnIgnored Whether or not a report should be made if
-     *                                    the file is ignored
+     * @param {boolean} isDirectPath True if the file
+     * was provided as a direct path (as opposed to being resolved from a glob)
      * @param {IgnoredPaths} ignoredPaths An instance of IgnoredPaths
      * @returns {void}
      */
-    function addFile(filename, shouldWarnIgnored, ignoredPaths) {
-        let ignored = false;
-        let isSilentlyIgnored;
-
-        if (ignoredPaths.contains(filename, "default")) {
-            ignored = (options.ignore !== false) && shouldWarnIgnored;
-            isSilentlyIgnored = !shouldWarnIgnored;
+    function addFile(filename, isDirectPath, ignoredPaths) {
+        if (addedFilenames.has(filename)) {
+            return;
         }
 
-        if (options.ignore !== false) {
-            if (ignoredPaths.contains(filename, "custom")) {
-                if (shouldWarnIgnored) {
-                    ignored = true;
-                } else {
-                    isSilentlyIgnored = true;
-                }
-            }
-        }
+        const shouldProcessCustomIgnores = options.ignore !== false;
+        const shouldLintIgnoredDirectPaths = options.ignore === false;
+        const fileMatchesIgnorePatterns = ignoredPaths.contains(filename, "default") ||
+            shouldProcessCustomIgnores && ignoredPaths.contains(filename, "custom");
 
-        if (isSilentlyIgnored && !ignored) {
-            return;
+        if (fileMatchesIgnorePatterns && isDirectPath && !shouldLintIgnoredDirectPaths) {
+            files.push({ filename, ignored: true });
+            addedFilenames.add(filename);
         }
 
-        if (added[filename]) {
-            return;
+        if (!fileMatchesIgnorePatterns || isDirectPath && shouldLintIgnoredDirectPaths) {
```

Comment:
I'd love to see parentheses here (around the && expression) for readability, as well.